src/mesh/mesh_function.C

```diff
@@ -556,19 +590,55 @@ void MeshFunction::discontinuous_gradient (const Point & p,
 
           const FEType & fe_type = this->_dof_map.variable_type(var);
 
-          std::unique_ptr<FEBase> point_fe (FEBase::build(dim, fe_type));
-          const std::vector<std::vector<RealGradient>> & dphi = point_fe->get_dphi();
-          point_fe->reinit(element, &point_list);
-
           // where the solution values for the var-th variable are stored
           std::vector<dof_id_type> dof_indices;
           this->_dof_map.dof_indices (element, dof_indices, var);
 
           Gradient grad(0.);
 
-          for (auto i : index_range(dof_indices))
-            grad.add_scaled(dphi[i][0], this->_vector(dof_indices[i]));
+          // for performance-reasons, we use different algorithms now.
+          // TODO: Check that both give the same result for finite elements.
+          // Otherwive it is wrong...
+#ifdef LIBMESH_ENABLE_INFINITE_ELEMENTS
+          if (!element->infinite())
+            {
+#endif
+              std::unique_ptr<FEBase> point_fe (FEBase::build(dim, fe_type));
+              const std::vector<std::vector<RealGradient>> & dphi = point_fe->get_dphi();
+              point_fe->reinit(element, & point_list);
+
+              temp_output[index] = grad;
```

Comment:
Should this be here?  It looks like it's redundantly zeroing out something we'll write the *correct* value to later.

Reply:
you are right; this is not necessary at all.